### PR TITLE
replace aho corasick 

### PR DESCRIPTION
**Pull request by jptosso**
This is a huge improvement, memory consumption is 60% less, and execution time was 233% faster

## Benchmark results using CRS:

**Old results:**
```
Preparing CRS...
CRS PATH: /var/folders/tf/23czq0cd4wd5v54zcvctvtzw0000gn/T/crs1564927737
goos: darwin
goarch: arm64
pkg: github.com/corazawaf/coraza/v3/testing
BenchmarkCRSCompilation-10             3         378610667 ns/op
BenchmarkCRSSimpleGET-10             613           1791376 ns/op
BenchmarkCRSSimplePOST-10            308           3499975 ns/op
PASS
ok      github.com/corazawaf/coraza/v3/testing  13.985s
```

**New results:**
```
Preparing CRS...
CRS PATH: /var/folders/tf/23czq0cd4wd5v54zcvctvtzw0000gn/T/crs1721984072
goos: darwin
goarch: arm64
pkg: github.com/corazawaf/coraza/v3/testing
BenchmarkCRSCompilation-10            21          48269002 ns/op
BenchmarkCRSSimpleGET-10             777           1396721 ns/op
BenchmarkCRSSimplePOST-10            489           2396956 ns/op
PASS
ok      github.com/corazawaf/coraza/v3/testing  5.969s
```

### CODE REVIEW

**Review by fzipi — APPROVED** [review]
Awesome results!

**Review by piyushroshan — APPROVED** [review]

File: operators/pm.go [hunk 1]
```diff
@@ -17,38 +17,47 @@ package operators
...
 func (o *pm) Evaluate(tx *coraza.Transaction, value string) bool {
-	value = strings.ToLower(value)
-	matches := o.matcher.MatchThreadSafe([]byte(value))
-	for i := 0; i < len(matches); i++ {
-		if i == 10 {
-			return true
+	if tx.Capture {
+		matches := o.matcher.FindAll(value)
+		for i, match := range matches {
+			if i == 10 {
```
Comment (jcchavezs) [inline operators/pm.go:50]: I know it isn't part of the PR but it would be really cool to get a hint on what this 10 means.

Comment (fzipi) [reply operators/pm.go:50]: Excellent question :) This is another constant that we should be naming properly.

10 is the maximum captures you can have (originally in modsecurity). This comes from https://github.com/SpiderLabs/ModSecurity/wiki/Reference-Manual-%28v2.x%29#capture
```
Up to 10 captures will be copied on a successful pattern match, each with a name consisting of a digit from 0 to 9. The TX.0 variable always contains the entire area that the regular expression matched. All the other variables contain the captured values, in the order in which the capturing parentheses appear in the regular expression. 
```

**Comment by jcchavezs** [comment]
@anuraaga worth for you to have a look at github.com/petar-dambovaliev/aho-corasick repository to see if all effort towards performance is enough?

**Comment by jptosso** [comment]
> @anuraaga worth for you to have a look at github.com/petar-dambovaliev/aho-corasick repository to see if all effort towards performance is enough?

Looks quite similar to our implementation, we would have to run benchmarks.

**Pull request merged**